test/blackbox/common/DDSBlackboxTestsMonitorService.cpp

```diff
@@ -2245,3 +2294,141 @@ TEST(DDSMonitorServiceTest, monitor_service_advanced_multiple_late_joiners)
 #endif //FASTDDS_STATISTICS
 }
 
+/**
+ * Regression test for Redmine issue #20625.
+ *
+ * The monitor service writer must correctly handle instances, removing old changes
+ * when its going to publish a new change in the same instance since it is
+ * defined as KEEP_LAST 1.
+ */
+TEST(DDSMonitorServiceTest,  monitor_service_properly_handles_topic_instances)
+{
+#ifdef FASTDDS_STATISTICS
+
+    //! In this test we do not need to enforce the validation of samples
+    std::bitset<statistics::STATUSES_SIZE> validation_mask;
+
+    //! Setup
+    MonitorServiceConsumer MSC(validation_mask);
+
+    MSC.init_monitor_service_reader();
+
+    //! Setup
+    MonitorServiceParticipant MSP;
+    auto test_transport = std::make_shared<test_UDPv4TransportDescriptor>();
+
+    std::atomic<uint8_t> n_gap_messages{0};
+
+    test_transport->drop_gap_messages_filter_ = [&n_gap_messages](CDRMessage_t& msg)
+            {
+
+                // Jump the reader entity id
+                msg.pos += 4;
+                // Read the writer's entity id
+                eprosima::fastrtps::rtps::EntityId_t writer_entity_id;
+                eprosima::fastrtps::rtps::CDRMessage::readEntityId(&msg, &writer_entity_id);
+
+                if (ENTITYID_MONITOR_SERVICE_WRITER == writer_entity_id)
+                {
+                    n_gap_messages.fetch_add(1);
+                }
+
+                return false;
+            };
+
+    DomainParticipantQos participant_qos;
+    participant_qos.transport().user_transports.push_back(test_transport);
+    participant_qos.transport().use_builtin_transports = false;
+
+    MSP.setup(participant_qos);
+    MSP.enable_monitor_service();
+
+    DataReaderQos dr_qos;
+    DataWriterQos dw_qos;
+
+    //! Set deadline as 1 sec
+    dr_qos.deadline().period = eprosima::fastrtps::Time_t{1, 000000000};
+    dw_qos.deadline().period = eprosima::fastrtps::Time_t{1, 000000000};
+
+    MSP.create_and_add_reader(dr_qos);
+    MSP.create_and_add_writer(dw_qos);
+
+    MSC.start_reception(std::list<MonitorServiceType::type>());
+
+    auto samples = default_helloworld_data_generator(4);
+    // Default hb period is 3 secs.
+    // Ensure that the monitor service writer sends at least one
+    MSP.send(samples, 1150);
+
+    ASSERT_TRUE(n_gap_messages > 0u);
+#endif //FASTDDS_STATISTICS
+}
+
+/**
+ * Regression test for Redmine issue #20625.
+ *
+ * A monitor service consumer receives only the last instance status when late joins.
+ */
```

Comment:
Is this test creating more instances? I would create different instances with keys and change the status of instance_1 (Ex. deadline_miss) and then change the status of instance_2 (Ex. liveliness_lost), then I would check if the last status of instance_1 and instance_2 are the desired ones, in this way we check that each history queue is managing correctly

Reply:
Not exactly. Previously, the monitor service was keeping in the history the different updates to the same instance (for example, the `DeadlineMissed` of the user-defined reader or writer) instead or removing the last change belonging to the instance when a new one was going to be published (as it is defined as `KEEP_LAST` `1`). 
The test intentionally produces several updates to the same instances and it is checked that a `GAP` message in sent from the writer to the reader to state that not all changes (by sequence number) on the writer are available. This is the desired behavior since it will mean that an old change for the same instance has been removed in favor of a new one.

Reply:
Thank you everything clear. Then I would only ask to add few more comments in the tests, please